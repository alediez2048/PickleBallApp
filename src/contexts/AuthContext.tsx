import React, {
  createContext,
  useContext,
  useState,
  useEffect,
  useCallback,
} from "react";
import { router } from "expo-router";
import {
  mockApi,
  FirstTimeProfileData,
  UpdateProfileData,
} from "@/services/mockApi";
import { storage } from "@/services/storage";
import { socialAuth } from "@/services/socialAuth";
import { Alert, Platform } from "react-native";
import { MembershipPlan } from "@/types/membership";
import {
  signUpWithEmail,
  signInWithEmail,
  getSession,
  logout,
} from "@/services/authService";
import { createUserProfile } from "@/services/userService";
import AsyncStorage from "@react-native-async-storage/async-storage";

interface PaymentMethod {
  id: string;
  last4: string;
  brand: string;
  expiryMonth: string;
  expiryYear: string;
  isDefault: boolean;
}

interface UserProfile {
  id?: string;
  email?: string;
  name?: string;
  isVerified?: boolean;
  skillLevel?: string;
  profileImage?:
    | string
    | {
        uri: string;
        base64: string;
        timestamp: number;
      };
  phoneNumber?: string;
  dateOfBirth?: string;
  address?: {
    street?: string;
    city?: string;
    state?: string;
    zipCode?: string;
    country?: string;
  };
  membership?: MembershipPlan;
  paymentMethods?: PaymentMethod[];
  hasCompletedProfile?: boolean;
}

interface AuthState {
  user: UserProfile | null;
  token: string | null;
  session: any | null;
  isLoading: boolean;
}

interface AuthContextType extends AuthState {
  // Authentication methods
  signIn: (email: string, password: string) => Promise<void>;
  signUp: (email: string, password: string, name: string) => Promise<void>;
  signOut: () => Promise<void>;

  // Social authentication methods
  signInWithGoogle: () => Promise<void>;
  signInWithFacebook: () => Promise<void>;

  // Profile management methods
  updateProfile: (updates: Partial<UserProfile>) => Promise<void>;
  updateFirstTimeProfile: (data: FirstTimeProfileData) => Promise<void>;
  updateMembership: (plan: MembershipPlan) => Promise<void>;
  updatePaymentMethods: (methods: PaymentMethod[]) => Promise<void>;
  updatePaymentMethod?: (hasPaymentMethod: boolean) => Promise<void>;

  // Auth state
  isAuthenticated: boolean;
}

const AuthContext = createContext<AuthContextType | undefined>(undefined);

export function AuthProvider({ children }: { children: React.ReactNode }) {
  const [state, setState] = useState<AuthState>({
    user: null,
    token: null,
    isLoading: true,
  });

  useEffect(() => {
    // Load token and user data from storage on app start
    loadStoredAuth();
  }, []);

  useEffect(() => {
    // Only navigate when auth state changes, not on every loading state change
    if (!state.isLoading && state.token === null) {
      router.replace("/login");
    }
  }, [state.token]);

  const loadStoredAuth = async () => {
    try {
      console.log("Loading stored auth data...");

      const [token, userString] = await Promise.all([
        AsyncStorage.getItem("auth_token"),
        AsyncStorage.getItem("user"),
      ]);

      console.log("Stored auth data:", { token, userString });

      if (token && userString) {
        const user = JSON.parse(userString);

        const { data, error } = await getSession();

        if (error || !data?.session) {
          console.warn("Session expired or invalid in Supabase");
          await AsyncStorage.multiRemove(["auth_token", "user"]);
          setState({
            token: null,
            session: null,
            user: null,
            isLoading: false,
          });
          return;
        }

        console.log("Supabase session is valid");
        setState({
          token: data.session.access_token,
          session: data.session,
          user,
          isLoading: false,
        });
      } else {
        setState({ token: null, session: null, user: null, isLoading: false });
      }
    } catch (error) {
      console.error("Error loading auth data:", error);
      setState({ token: null, session: null, user: null, isLoading: false });
    }
  };

  const signUp = async (email: string, password: string, name: string) => {
    try {
      console.log("[Auth] Starting registration process", { email, name });
      setState((prev) => ({ ...prev, isLoading: true }));

      const { data, error } = await signUpWithEmail(email, password);
      if (error) throw new Error(error.message);

      const userId = data?.user?.id;
      if (userId) {
        await createUserProfile(userId, { name, email });
        setState({
          session: data.session,
          token: data.session?.access_token || null,
          user: data.user,
          isLoading: false,
        });

        // Store the auth data
        await Promise.all([
          data.session?.access_token
            ? storage.setItem("auth_token", data.session.access_token)
            : Promise.resolve(),
          storage.setItem("user", JSON.stringify(data.user)),
        ]);
        console.log("[Auth] User data stored in local storage");
      }
    } catch (error) {
      console.error("[Auth] Registration failed:", error);
      setState((prev) => ({ ...prev, isLoading: false }));
      throw error;
    }
  };

  const signIn = async (email: string, password: string) => {
    try {
      console.log("Signing in...", { email });
      setState((prev) => ({ ...prev, isLoading: true }));

      const { data, error } = await signInWithEmail(email, password);

      if (error || !data.session) {
        throw new Error(error?.message || "No session returned");
      }

      const token = data.session.access_token;
      const user = data.user;
      const session = data.session;

      console.log("Sign in successful:", { token, user });

      await Promise.all([
        AsyncStorage.setItem("auth_token", token),
        AsyncStorage.setItem("user", JSON.stringify(user)),
      ]);

      console.log("Auth data stored");

      setState({ token, session, user, isLoading: false });
    } catch (error: any) {
      console.error("Sign in error:", error.message);
      setState((prev) => ({ ...prev, isLoading: false }));
      throw error;
    }
  };

  const signOut = async () => {
    try {
      console.log("Signing out...");
      setState((prev) => ({ ...prev, isLoading: true }));

      await Promise.all([
        storage.removeItem("auth_token"),
        storage.removeItem("user"),
      ]);
      console.log("Auth data removed");

      setState({ token: null, user: null, isLoading: false });
    } catch (error) {
      console.error("Sign out error:", error);
      setState((prev) => ({ ...prev, isLoading: false }));
      throw error;
    }
  };

  const handleSocialAuthSuccess = async (
    provider: "google" | "facebook",
    token: string,
    user: { id: string; email: string; name: string; photoUrl?: string }
  ) => {
    try {
      const response = await mockApi.socialAuth({ token, provider, user });

      await Promise.all([
        storage.setItem("auth_token", response.token),
        storage.setItem("user", JSON.stringify(response.user)),
      ]);

      setState({
        token: response.token,
        user: response.user,
        isLoading: false,
      });
    } catch (error) {
      console.error(`${provider} auth error:`, error);
      Alert.alert("Error", `Failed to authenticate with ${provider}`);
    }
  };

  const signInWithGoogle = async () => {
    try {
      setState((prev) => ({ ...prev, isLoading: true }));
      const result = await socialAuth.signInWithGoogle();

      if (result.type === "success" && result.token && result.user) {
        await handleSocialAuthSuccess("google", result.token, result.user);
      } else {
        throw new Error(result.message || "Google sign in failed");
      }
    } catch (error) {
      console.error("Google sign in error:", error);
      Alert.alert("Error", "Failed to sign in with Google");
      setState((prev) => ({ ...prev, isLoading: false }));
    }
  };

  const signInWithFacebook = async () => {
    try {
      setState((prev) => ({ ...prev, isLoading: true }));
      const result = await socialAuth.signInWithFacebook();

      if (result.type === "success" && result.token && result.user) {
        await handleSocialAuthSuccess("facebook", result.token, result.user);
      } else {
        throw new Error(result.message || "Facebook sign in failed");
      }
    } catch (error) {
      console.error("Facebook sign in error:", error);
      Alert.alert("Error", "Failed to sign in with Facebook");
      setState((prev) => ({ ...prev, isLoading: false }));
    }
  };

  const updateProfile = async (updates: Partial<UserProfile>) => {
    try {
      setState((prev) => ({ ...prev, isLoading: true }));

      if (!state.user?.email) {
        throw new Error("No authenticated user");
      }

      // Map UserProfile to UpdateProfileData
      const profileUpdates: UpdateProfileData = { ...updates } as any;

      // Handle address mapping if it exists
      if (updates.address) {
        profileUpdates.address = {
          address: updates.address.street || "",
          city: updates.address.city || "",
          state: updates.address.state || "",
          zipCode: updates.address.zipCode || "",
          country: updates.address.country || "United States",
        };
      }

      const { user: updatedUser } = await mockApi.updateProfile(
        state.user.email,
        profileUpdates
      );

      // Store the updated user data
      await storage.setItem("user", JSON.stringify(updatedUser));

      setState((prev) => ({
        ...prev,
        user: updatedUser,
        isLoading: false,
      }));

      console.log("Profile updated successfully:", updatedUser);
    } catch (error) {
      console.error("Profile update error:", error);
      setState((prev) => ({ ...prev, isLoading: false }));
      throw error;
    }
  };

  const updateFirstTimeProfile = async (data: FirstTimeProfileData) => {
    try {
      console.debug("[AuthContext] Starting first time profile update", {
        platform: Platform.OS,
        hasUser: !!state.user,
        userEmail: state.user?.email,
        data,
      });

      setState((prev) => ({ ...prev, isLoading: true }));

      if (!state.user?.email) {
        console.error("[AuthContext] No authenticated user found");
        throw new Error("No authenticated user");
      }

      // Map FirstTimeProfileData to UpdateProfileData
      const profileData: UpdateProfileData = {
        skillLevel: data.skillLevel,
        displayName: data.displayName,
        phoneNumber: data.phoneNumber,
        dateOfBirth: data.dateOfBirth,
        address: {
          address: data.address.address,
          city: data.address.city,
          state: data.address.state,
          zipCode: data.address.zipCode,
          country: data.address.country,
        },
        hasCompletedProfile: true,
      };

      console.debug("[AuthContext] Calling mockApi.updateProfile", {
        email: state.user.email,
        profileData,
      });

      const { user: updatedUser } = await mockApi.updateProfile(
        state.user.email,
        profileData
      );

      console.debug("[AuthContext] Profile update API call successful", {
        updatedUser,
      });

      // After successful profile update, update the stored user data
      const userWithProfile = {
        ...updatedUser,
        hasCompletedProfile: true,
      };

      console.debug("[AuthContext] Storing updated user data");
      await storage.setItem("user", JSON.stringify(userWithProfile));

      console.debug("[AuthContext] Updating state with new user data");
      setState((prev) => ({
        ...prev,
        user: userWithProfile,
        isLoading: false,
      }));

      console.debug(
        "[AuthContext] First time profile update completed successfully"
      );
    } catch (error) {
      console.error("[AuthContext] Profile update error:", error);
      setState((prev) => ({ ...prev, isLoading: false }));
      throw error;
    }
  };

  const updateMembership = async (plan: MembershipPlan) => {
    try {
      if (!state.user) {
        throw new Error("No authenticated user");
      }

      // Update user with new membership plan
      const updatedUser = {
        ...state.user,
        membership: plan,
      };

      // Store the updated user data
      await storage.setItem("user", JSON.stringify(updatedUser));

      // Update state
      setState((prev) => ({
        ...prev,
        user: updatedUser,
      }));

      console.log("Membership updated successfully:", plan);
    } catch (error) {
      console.error("Error updating membership:", error);
      throw error;
    }
  };

  const updatePaymentMethods = async (methods: PaymentMethod[]) => {
    try {
      if (!state.user) {
        throw new Error("No authenticated user");
      }

      // Update user with new payment methods
      const updatedUser = {
        ...state.user,
        paymentMethods: methods,
      };

      // Store the updated user data
      await storage.setItem("user", JSON.stringify(updatedUser));

      // Update state
      setState((prev) => ({
        ...prev,
        user: updatedUser,
      }));

      console.log("Payment methods updated successfully:", methods);
    } catch (error) {
      console.error("Error updating payment methods:", error);
      throw error;
    }
  };

  const updatePaymentMethod = async (hasPaymentMethod: boolean) => {
    try {
      if (!state.user) {
        throw new Error("No authenticated user");
      }

      // Update user with new payment method
      const updatedUser = {
        ...state.user,
        paymentMethods: hasPaymentMethod
          ? [
              {
                id: "new_payment_method",
                last4: "XXXX",
                brand: "New",
                expiryMonth: "12",
                expiryYear: "2024",
                isDefault: true,
              },
            ]
          : [],
      };

      // Store the updated user data
      await storage.setItem("user", JSON.stringify(updatedUser));

      // Update state
      setState((prev) => ({
        ...prev,
        user: updatedUser,
      }));

      console.log(
        "Payment method updated successfully:",
        hasPaymentMethod ? "Added" : "Removed"
      );
    } catch (error) {
      console.error("Error updating payment method:", error);
      throw error;
    }
  };

  const value = {
    ...state,
    signIn,
    signUp,
    signOut,
    signInWithGoogle,
    signInWithFacebook,
    updateProfile,
    updateFirstTimeProfile,
    updateMembership,
    updatePaymentMethods,
    updatePaymentMethod,
    isAuthenticated: !!state.token,
  };

  return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>;
}

export function useAuth() {
  const context = useContext(AuthContext);
  if (context === undefined) {
    throw new Error("useAuth must be used within an AuthProvider");
  }
  return context;
}
