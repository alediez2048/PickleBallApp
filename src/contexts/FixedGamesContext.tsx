// Context for Fixed Games CRUD operations
import React, { createContext, useContext, useState, useCallback } from "react";
import {
  getFixedGames,
  getFixedGame,
  createFixedGame,
  updateFixedGame,
  deleteFixedGame,
} from "@/services/fixedGamesService";
import { FixedGame, FixedGameInput } from "@/types/fixedGames";

interface FixedGamesContextProps {
  fixedGames: FixedGame[];
  loading: boolean;
  fetchFixedGames: () => Promise<void>;
  getFixedGame: (id: string) => Promise<FixedGame | null>;
  createFixedGame: (input: FixedGameInput) => Promise<FixedGame | null>;
  updateFixedGame: (
    id: string,
    input: Partial<FixedGameInput>
  ) => Promise<boolean>;
  deleteFixedGame: (id: string) => Promise<boolean>;
}

const FixedGamesContext = createContext<FixedGamesContextProps | undefined>(
  undefined
);

export const FixedGamesProvider: React.FC<{ children: React.ReactNode }> = ({
  children,
}) => {
  const [fixedGames, setFixedGames] = useState<FixedGame[]>([]);
  const [loading, setLoading] = useState(false);

  const fetchFixedGames = useCallback(async () => {
    setLoading(true);
    try {
      const data = await getFixedGames();
      setFixedGames(data);
    } finally {
      setLoading(false);
    }
  }, []);

  const getFixedGameById = async (id: string) => {
    return await getFixedGame(id);
  };

  const create = async (input: FixedGameInput) => {
    setLoading(true);
    try {
      const created = await createFixedGame(input);
      if (created) await fetchFixedGames();
      return created;
    } finally {
      setLoading(false);
    }
  };

  const update = async (id: string, input: Partial<FixedGameInput>) => {
    setLoading(true);
    try {
      const ok = await updateFixedGame(id, input);
      if (ok) await fetchFixedGames();
      return ok;
    } finally {
      setLoading(false);
    }
  };

  const remove = async (id: string) => {
    setLoading(true);
    try {
      const ok = await deleteFixedGame(id);
      if (ok) await fetchFixedGames();
      return ok;
    } finally {
      setLoading(false);
    }
  };

  return (
    <FixedGamesContext.Provider
      value={{
        fixedGames,
        loading,
        fetchFixedGames,
        getFixedGame: getFixedGameById,
        createFixedGame: create,
        updateFixedGame: update,
        deleteFixedGame: remove,
      }}
    >
      {children}
    </FixedGamesContext.Provider>
  );
};

export function useFixedGames() {
  const ctx = useContext(FixedGamesContext);
  if (!ctx)
    throw new Error("useFixedGames must be used within FixedGamesProvider");
  return ctx;
}
