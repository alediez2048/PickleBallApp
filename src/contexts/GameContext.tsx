import React, { createContext, useContext, useReducer, useCallback, useEffect } from 'react';
import { Game, SkillLevel } from '@/types/game';
import { gamesApi, GameFilters } from '@/services/api/games';
import { ApiError } from '@/services/api/client';

// State interface
interface GameState {
  games: Game[];
  loading: boolean;
  error: string | null;
  optimisticUpdates: Map<string, Game>;
  prefetchedGames: Map<string, Game>;
}

// Action types
type GameAction =
  | { type: 'SET_LOADING'; payload: boolean }
  | { type: 'SET_ERROR'; payload: string | null }
  | { type: 'SET_GAMES'; payload: Game[] }
  | { type: 'ADD_GAME'; payload: Game }
  | { type: 'UPDATE_GAME'; payload: Game }
  | { type: 'DELETE_GAME'; payload: string }
  | { type: 'SET_OPTIMISTIC_UPDATE'; payload: { id: string; game: Game } }
  | { type: 'CLEAR_OPTIMISTIC_UPDATE'; payload: string }
  | { type: 'SET_PREFETCHED_GAME'; payload: { id: string; game: Game } }
  | { type: 'CLEAR_PREFETCHED_GAME'; payload: string };

// Initial state
const initialState: GameState = {
  games: [],
  loading: false,
  error: null,
  optimisticUpdates: new Map(),
  prefetchedGames: new Map(),
};

// Reducer
function gameReducer(state: GameState, action: GameAction): GameState {
  switch (action.type) {
    case 'SET_LOADING':
      return { ...state, loading: action.payload };
    case 'SET_ERROR':
      return { ...state, error: action.payload };
    case 'SET_GAMES':
      return { ...state, games: action.payload };
    case 'ADD_GAME':
      return { ...state, games: [...state.games, action.payload] };
    case 'UPDATE_GAME': {
      const newGames = state.games.map(game => 
        game.id === action.payload.id ? action.payload : game
      );
      const newOptimisticUpdates = new Map(state.optimisticUpdates);
      newOptimisticUpdates.delete(action.payload.id);
      return {
        ...state,
        games: newGames,
        optimisticUpdates: newOptimisticUpdates,
      };
    }
    case 'DELETE_GAME': {
      const newGames = state.games.filter(game => game.id !== action.payload);
      const newOptimisticUpdates = new Map(state.optimisticUpdates);
      newOptimisticUpdates.delete(action.payload);
      return {
        ...state,
        games: newGames,
        optimisticUpdates: newOptimisticUpdates,
      };
    }
    case 'SET_OPTIMISTIC_UPDATE': {
      const newOptimisticUpdates = new Map(state.optimisticUpdates);
      newOptimisticUpdates.set(action.payload.id, action.payload.game);
      return {
        ...state,
        optimisticUpdates: newOptimisticUpdates,
      };
    }
    case 'CLEAR_OPTIMISTIC_UPDATE': {
      const newOptimisticUpdates = new Map(state.optimisticUpdates);
      newOptimisticUpdates.delete(action.payload);
      return {
        ...state,
        optimisticUpdates: newOptimisticUpdates,
      };
    }
    case 'SET_PREFETCHED_GAME': {
      const newPrefetchedGames = new Map(state.prefetchedGames);
      newPrefetchedGames.set(action.payload.id, action.payload.game);
      return {
        ...state,
        prefetchedGames: newPrefetchedGames,
      };
    }
    case 'CLEAR_PREFETCHED_GAME': {
      const newPrefetchedGames = new Map(state.prefetchedGames);
      newPrefetchedGames.delete(action.payload);
      return {
        ...state,
        prefetchedGames: newPrefetchedGames,
      };
    }
    default:
      return state;
  }
}

// Context interface
interface GameContextType extends Omit<GameState, 'optimisticUpdates' | 'prefetchedGames'> {
  fetchGames: (filters?: GameFilters) => Promise<void>;
  createGame: (gameData: Omit<Game, 'id' | 'status'>) => Promise<void>;
  updateGame: (gameId: string, gameData: Partial<Game>) => Promise<void>;
  deleteGame: (gameId: string) => Promise<void>;
  joinGame: (gameId: string) => Promise<void>;
  leaveGame: (gameId: string) => Promise<void>;
  prefetchGame: (gameId: string) => Promise<void>;
  getGame: (gameId: string) => Game | undefined;
}

// Create context
const GameContext = createContext<GameContextType | undefined>(undefined);

// Provider component
export function GameProvider({ children }: { children: React.ReactNode }) {
  const [state, dispatch] = useReducer(gameReducer, initialState);

  const handleError = (error: unknown) => {
    const message = error instanceof ApiError 
      ? error.message 
      : 'An unexpected error occurred';
    dispatch({ type: 'SET_ERROR', payload: message });
  };

  const getGame = useCallback((gameId: string): Game | undefined => {
    // Check optimistic updates first
    const optimisticGame = state.optimisticUpdates.get(gameId);
    if (optimisticGame) return optimisticGame;

    // Then check prefetched games
    const prefetchedGame = state.prefetchedGames.get(gameId);
    if (prefetchedGame) return prefetchedGame;

    // Finally check regular games
    return state.games.find(game => game.id === gameId);
  }, [state.games, state.optimisticUpdates, state.prefetchedGames]);

  const prefetchGame = useCallback(async (gameId: string) => {
    try {
      const response = await gamesApi.getGame(gameId);
      dispatch({ 
        type: 'SET_PREFETCHED_GAME', 
        payload: { id: gameId, game: response.data } 
      });
    } catch (error) {
      console.error('Error prefetching game:', error);
    }
  }, []);

  const fetchGames = useCallback(async (filters?: GameFilters) => {
    dispatch({ type: 'SET_LOADING', payload: true });
    dispatch({ type: 'SET_ERROR', payload: null });
    try {
      const response = await gamesApi.getGames(filters);
      dispatch({ type: 'SET_GAMES', payload: response.data });
    } catch (error) {
      handleError(error);
    } finally {
      dispatch({ type: 'SET_LOADING', payload: false });
    }
  }, []);

  const createGame = useCallback(async (gameData: Omit<Game, 'id' | 'status'>) => {
    dispatch({ type: 'SET_LOADING', payload: true });
    dispatch({ type: 'SET_ERROR', payload: null });
    try {
      const response = await gamesApi.createGame(gameData);
      dispatch({ type: 'ADD_GAME', payload: response.data });
    } catch (error) {
      handleError(error);
      throw error;
    } finally {
      dispatch({ type: 'SET_LOADING', payload: false });
    }
  }, []);

  const updateGame = useCallback(async (gameId: string, gameData: Partial<Game>) => {
    const currentGame = getGame(gameId);
    if (!currentGame) throw new Error('Game not found');

    // Apply optimistic update
    const optimisticGame = { ...currentGame, ...gameData };
    dispatch({ 
      type: 'SET_OPTIMISTIC_UPDATE', 
      payload: { id: gameId, game: optimisticGame } 
    });

    try {
      const response = await gamesApi.updateGame(gameId, gameData);
      dispatch({ type: 'UPDATE_GAME', payload: response.data });
    } catch (error) {
      // Revert optimistic update
      dispatch({ type: 'CLEAR_OPTIMISTIC_UPDATE', payload: gameId });
      handleError(error);
      throw error;
    }
  }, [getGame]);

  const deleteGame = useCallback(async (gameId: string) => {
    // Optimistically remove the game
    const deletedGame = getGame(gameId);
    if (deletedGame) {
      dispatch({ 
        type: 'SET_OPTIMISTIC_UPDATE', 
        payload: { id: gameId, game: { ...deletedGame, status: 'Cancelled' } } 
      });
    }

    try {
      await gamesApi.deleteGame(gameId);
      dispatch({ type: 'DELETE_GAME', payload: gameId });
    } catch (error) {
      // Revert optimistic update
      dispatch({ type: 'CLEAR_OPTIMISTIC_UPDATE', payload: gameId });
      handleError(error);
      throw error;
    }
  }, [getGame]);

  const joinGame = useCallback(async (gameId: string) => {
    const currentGame = getGame(gameId);
    if (!currentGame) throw new Error('Game not found');

    // Apply optimistic update
    const optimisticGame = { 
      ...currentGame, 
      currentPlayers: currentGame.currentPlayers + 1 
    };
    dispatch({ 
      type: 'SET_OPTIMISTIC_UPDATE', 
      payload: { id: gameId, game: optimisticGame } 
    });

    try {
      const response = await gamesApi.joinGame(gameId);
      dispatch({ type: 'UPDATE_GAME', payload: response.data });
    } catch (error) {
      // Revert optimistic update
      dispatch({ type: 'CLEAR_OPTIMISTIC_UPDATE', payload: gameId });
      handleError(error);
      throw error;
    }
  }, [getGame]);

  const leaveGame = useCallback(async (gameId: string) => {
    const currentGame = getGame(gameId);
    if (!currentGame) throw new Error('Game not found');

    // Apply optimistic update
    const optimisticGame = { 
      ...currentGame, 
      currentPlayers: currentGame.currentPlayers - 1 
    };
    dispatch({ 
      type: 'SET_OPTIMISTIC_UPDATE', 
      payload: { id: gameId, game: optimisticGame } 
    });

    try {
      const response = await gamesApi.leaveGame(gameId);
      dispatch({ type: 'UPDATE_GAME', payload: response.data });
    } catch (error) {
      // Revert optimistic update
      dispatch({ type: 'CLEAR_OPTIMISTIC_UPDATE', payload: gameId });
      handleError(error);
      throw error;
    }
  }, [getGame]);

  const value = {
    games: state.games,
    loading: state.loading,
    error: state.error,
    fetchGames,
    createGame,
    updateGame,
    deleteGame,
    joinGame,
    leaveGame,
    prefetchGame,
    getGame,
  };

  return <GameContext.Provider value={value}>{children}</GameContext.Provider>;
}

// Hook for using the context
export function useGames() {
  const context = useContext(GameContext);
  if (context === undefined) {
    throw new Error('useGames must be used within a GameProvider');
  }
  return context;
} 