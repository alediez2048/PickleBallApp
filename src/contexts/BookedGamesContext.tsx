import React, { createContext, useContext, useState, useEffect } from "react";
import { useAuth } from "@/contexts/AuthContext";
import {
  createBookedGame as createBookedGameService,
  listBookedGames as listBookedGamesService,
  updateBookedGame as updateBookedGameService,
} from "@services/bookedGamesService";
import { getGame, updateGame } from "@services/gamesService";

import { BookedGame } from "@/types/bookedGames";
import { User } from "@/types/games";

interface BookedGamesContextType {
  bookedGames: BookedGame[];
  isLoading: boolean;
  error: Error | null;
  refreshBookedGames: () => Promise<void>;
  addBookedGame: (game: Omit<BookedGame, "id">) => Promise<BookedGame | null>;
  cancelBooking: (gameId: string) => Promise<void>;
  clearAllGames: () => Promise<void>;
  listBookedGamesForUser: () => Promise<BookedGame[]>;
}

const BookedGamesContext = createContext<BookedGamesContextType | undefined>(
  undefined
);

export const BookedGamesProvider: React.FC<{ children: React.ReactNode }> = ({
  children,
}) => {
  const [bookedGames, setBookedGames] = useState<BookedGame[]>([]);
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState<Error | null>(null);
  const { user } = useAuth();

  const refreshBookedGames = async () => {
    if (!user?.email) return;
    setIsLoading(true);
    setError(null);
    try {
      const { data, error: fetchError } = await listBookedGamesService();
      if (fetchError) throw fetchError;
      setBookedGames(data || []);
    } catch (err) {
      setError(
        err instanceof Error ? err : new Error("Failed to fetch booked games")
      );
    } finally {
      setIsLoading(false);
    }
  };

  const addBookedGame = async (game: Omit<BookedGame, "id">) => {
    if (!user?.email) throw new Error("User not authenticated");
    try {
      // Fetch the current game details first
      const currentGameData = await getGame(game.game_id);
      if (!currentGameData) throw new Error("Game not found");

      // Prepare the new player object using User type
      const newPlayer: User = {
        id: user.id,
        email: user.email,
        name: user.display_name || user.name || "",
        skill_level: currentGameData.skill_level,
        phone_number: user.phone_number || "",
      };

      // Prepare the updated players array
      const updatedPlayers: User[] = [
        ...(currentGameData.players || []),
        newPlayer,
      ];

      // Register the booked game
      const bookedGame = await createBookedGameService({ ...game });
      if (bookedGame) {
        await updateGame(bookedGame.game_id, {
          registered_count: updatedPlayers.length,
          players: updatedPlayers,
        });

        setBookedGames((prev) => [bookedGame, ...prev]);
        return bookedGame;
      }
      return null;
    } catch (err) {
      console.error("Error adding booked game:", err);
      throw err instanceof Error ? err : new Error("Failed to book game");
    }
  };

  const cancelBooking = async (gameId: string) => {
    if (!user?.email) throw new Error("User not authenticated");
    try {
      await updateBookedGameService(gameId, { status: "cancelled" });
      setBookedGames((prev) =>
        prev.map((game) =>
          game.id === gameId ? { ...game, status: "cancelled" as const } : game
        )
      );
    } catch (err) {
      throw err instanceof Error ? err : new Error("Failed to cancel booking");
    }
  };

  const clearAllGames = async () => {
    if (!user?.email) return;
    setIsLoading(true);
    try {
      setBookedGames([]);
    } catch (err) {
      throw err instanceof Error ? err : new Error("Failed to clear games");
    } finally {
      setIsLoading(false);
    }
  };

  /**
   * Fetch upcoming booked games for the current user in the next 7 days
   */
  const listBookedGamesForUser = async () => {
    if (!user?.id) return [];
    const today = new Date();
    const startDate = today.toISOString().split("T")[0];
    const endDateObj = new Date(today);
    endDateObj.setDate(today.getDate() + 7);
    const endDate = endDateObj.toISOString().split("T")[0];
    const { data, error } = await listBookedGamesService({
      userId: user.id,
      dateRange: { startDate, endDate },
      status: "upcoming",
    });
    if (error) {
      console.error("Error fetching upcoming booked games for user:", error);
      return [];
    }
    return data || [];
  };

  useEffect(() => {
    refreshBookedGames();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [user?.email]);

  return (
    <BookedGamesContext.Provider
      value={{
        bookedGames,
        isLoading,
        error,
        refreshBookedGames,
        addBookedGame,
        cancelBooking,
        clearAllGames,
        listBookedGamesForUser,
      }}
    >
      {children}
    </BookedGamesContext.Provider>
  );
};

export function useBookedGames() {
  const context = useContext(BookedGamesContext);
  if (context === undefined) {
    throw new Error("useBookedGames must be used within a BookedGamesProvider");
  }
  return context;
}
