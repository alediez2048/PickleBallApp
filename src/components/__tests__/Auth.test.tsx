import React from "react";
import renderer from "react-test-renderer";
import { View, Text, Pressable } from "react-native";
import { AuthProvider, useAuth } from "@/contexts/AuthContext";
import { mockApi } from "@/services/mockApi";

// Mock router
jest.mock("expo-router", () => ({
  router: {
    replace: jest.fn(),
    push: jest.fn(),
    back: jest.fn(),
  },
}));

// Test component that uses auth context
const TestComponent = () => {
  const { isAuthenticated, signIn } = useAuth();

  return (
    <View>
      <Text testID="auth-status">
        {isAuthenticated ? "logged-in" : "logged-out"}
      </Text>
      <Pressable
        testID="login-button"
        onPress={() => signIn("test@example.com", "password")}
      />
    </View>
  );
};

describe("Authentication Flow", () => {
  beforeEach(() => {
    // Clear all mocks before each test
    jest.clearAllMocks();
  });

  it("renders initial unauthenticated state", () => {
    // Mock the auth hook to return unauthenticated state
    jest.mock("@/contexts/AuthContext", () => ({
      AuthProvider: ({ children }: { children: React.ReactNode }) => children,
      useAuth: () => ({
        isAuthenticated: false,
        signIn: jest.fn(),
      }),
    }));
    
    const tree = renderer
      .create(<TestComponent />)
      .toJSON();
    
    expect(tree).toMatchSnapshot('unauthenticated state');
  });

  it("renders authenticated state", () => {
    // Mock the auth hook to return authenticated state
    jest.mock("@/contexts/AuthContext", () => ({
      AuthProvider: ({ children }: { children: React.ReactNode }) => children,
      useAuth: () => ({
        isAuthenticated: true,
        signIn: jest.fn(),
      }),
    }));
    
    const tree = renderer
      .create(<TestComponent />)
      .toJSON();
    
    expect(tree).toMatchSnapshot('authenticated state');
  });
  
  // Note: We can't easily test the state transitions using snapshot testing,
  // as that would require useEffect and state changes that affect snapshots.
  // This would be better tested with integration tests or using a testing
  // framework that can handle asynchronous state changes.
});
