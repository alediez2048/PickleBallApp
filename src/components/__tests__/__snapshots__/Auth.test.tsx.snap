// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authentication Flow renders authenticated state: authenticated state 1`] = `null`;

exports[`Authentication Flow renders initial unauthenticated state: unauthenticated state 1`] = `null`;
